### PR TITLE
Publish -SNAPSHOT (and release) builds to Maven

### PR DESCRIPTION
**Pull request by reta**
Signed-off-by: Andriy Redko <andriy.redko@aiven.io>

### Description
Publish -SNAPSHOT (and release) builds to Maven

### Issues Resolved
Closes https://github.com/opensearch-project/spring-data-opensearch/issues/9

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by peterzhuamazon** [comment]
Hi @reta when is this prepare to release?
Is this a blocker to 2.4.0?
If not, I think we can push it a bit after 2.4.0.

Also, could you open an issue on opensearch-project/opensearch-build so we can track this process?
And my teammate can also help if I am not available.

Thanks.

**Comment by reta** [comment]
Thanks @peterzhuamazon 

> Is this a blocker to 2.4.0?
> If not, I think we can push it a bit after 2.4.0.

No, it is not, we are targeting at least mid-November for release (but snapshots could be available earlier if possible)

> opensearch-project/opensearch-build

Done, https://github.com/opensearch-project/opensearch-build/issues/2846

**Comment by reta** [comment]
@peterzhuamazon any chance we could resume work on that? thank you!

**Comment by peterzhuamazon** [comment]
> @peterzhuamazon any chance we could resume work on that? thank you!

Hi @bbarani @gaiksaya we can probably help with this once the universal release mechanism is fully ready.

Thanks.

File: jenkins/publish-snapshot.jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+                    // https://docs.gradle.org/current/samples/sample_publishing_credentials.html
+                    // https://docs.gradle.org/current/userguide/build_environment.html#sec:project_properties
+                    withCredentials([usernamePassword(credentialsId: 'jenkins-sonatype-creds', usernameVariable: 'ORG_GRADLE_PROJECT_snapshotRepoUsername', passwordVariable: 'ORG_GRADLE_PROJECT_snapshotRepoPassword')]) {
+                        echo("Git checkout ${REPO_URL} on ${ref_final} for spring-data-opensearch release")
+                        checkout([
+                            $class: 'GitSCM', userRemoteConfigs: [[url: "${REPO_URL}" ]],
+                            branches: [[name: "${ref_final}"]]
+                        ])
+                        sh('./gradlew --no-daemon publishPublishMavenPublicationToSnapshotRepoRepository')
```
Comment (rursprung) [inline jenkins/publish-snapshot.jenkinsfile:64]: is `publishPublishMavenPublicationToSnapshotRepoRepository` really correct or a search&replace mistake?

Comment (reta) [reply jenkins/publish-snapshot.jenkinsfile:64]: > is `publishPublishMavenPublicationToSnapshotRepoRepository` really correct or a search&replace mistake?

It is really correct:

```
$ ./gradlew tasks | grep publishPublishMavenPublicationToSnapshotRepoRepository
publishPublishMavenPublicationToSnapshotRepoRepository - Publishes Maven publication 'publishMaven' to Maven repository 'snapshotRepo
```

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
@reta Merge at will.

**Comment by reta** [comment]
> @reta Merge at will.

Thanks a lot @dblock ! We sadly need a build team to configure webhooks and Jenkins, @bbarani @gaiksaya any updates on that guys?

**Comment by gaiksaya** [comment]
Hey! Tagging @rishabh6788 to check the ETA for the maven library. It will reduce the code to 1/4th as same set of steps are used for almost everything that is published on maven

**Comment by rishabh6788** [comment]
@reta @gaiksaya I should have an update by the end of this week.

**Comment by reta** [comment]
> @reta @gaiksaya I should have an update by the end of this week.

@rishabh6788 any updates? thanks

**Comment by reta** [comment]
@bbarani @peterzhuamazon could you please give me us a hand here? thank you.

**Comment by rishabh6788** [comment]
@reta sorry for the delay, got caught up in new release. Here is the PR we are working on to create a common library to auto-publish maven artifacts. Should get merged early next week. 
https://github.com/opensearch-project/opensearch-build-libraries/pull/83

**Comment by reta** [comment]
> @reta sorry for the delay, got caught up in new release. Here is the PR we are working on to create a common library to auto-publish maven artifacts. Should get merged early next week. [opensearch-project/opensearch-build-libraries#83](https://github.com/opensearch-project/opensearch-build-libraries/pull/83)

@rishabh6788 I see the pull request got merged, should I switch to `publishToMaven` instead? @peterzhuamazon what about web hooks? thank you guys.

**Comment by gaiksaya** [comment]
Hi @reta The publishing to snapshot feature is still to be integrated with the above library. Is it okay if we raise a PR and then close this one?

**Comment by reta** [comment]
> Hi @reta The publishing to snapshot feature is still to be integrated with the above library. Is it okay if we raise a PR and then close this one?

Sure, please go ahead @gaiksaya

**Comment by gaiksaya** [comment]
Quick question @reta : Is there any estimated date of release planned for this?

**Comment by reta** [comment]
> Quick question @reta : Is there any estimated date of release planned for this?

Yes, we planned for mid-November, 2022 :(

**Comment by gaiksaya** [comment]
> > Quick question @reta : Is there any estimated date of release planned for this?
> 
> @gaiksaya Yes, we planned for mid-November, 2022 :(

How about we add the release workflow first till snapshot feature gets added? Will that work or do you need both?

**Comment by reta** [comment]
> > > Quick question @reta : Is there any estimated date of release planned for this?
> > 
> > 
> > @gaiksaya Yes, we planned for mid-November, 2022 :(
> 
> How about we add the release workflow first till snapshot feature gets added? Will that work or do you need both?

Yes, it will work, thank you.

**Comment by gaiksaya** [comment]
>Yes, it will work, thank you.

Added release workflow here: https://github.com/opensearch-project/spring-data-opensearch/pull/74

**Comment by reta** [comment]
> > Yes, it will work, thank you.
> 
> Added release workflow here: #74

✔️ thank you @gaiksaya , we should close this pull request, right?

**Comment by gaiksaya** [comment]
> > > Yes, it will work, thank you.
> > 
> > 
> > Added release workflow here: #74
> 
> ✔️ thank you @gaiksaya , we should close this pull request, right?

Yes! I believe we can create another GH issue to track the publishing snapshots.

**Comment by reta** [comment]
> Yes! I believe we can create another GH issue to track the publishing snapshots.

We already have one: https://github.com/opensearch-project/spring-data-opensearch/issues/9

**Pull request closed without merging**